### PR TITLE
emscripten - LLVM to JavaScript compiler

### PR DESCRIPTION
**Pull request by timeraider4u**
added ebuilds for emscripten, see http://kripken.github.io/emscripten-site/docs/building_from_source/building_emscripten_from_source_on_linux.html and http://kripken.github.io/emscripten-site/docs/building_from_source/building_fastcomp_manually_from_source.html for detailed compilation instructions

### CODE REVIEW

File: dev-util/emscripten/emscripten-1.34.11.ebuild [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+DEST="/usr/share/"
+TEST="${WORKDIR}/test/"
+
+src_prepare() {
+	cp "${FILESDIR}/emscripten.config" "${S}/"
+	cp "${FILESDIR}/99emscripten" "${S}/"
+}
+
+src_test() {
```
Comment (marbre) [inline dev-util/emscripten/emscripten-1.34.11.ebuild]: If `src_test()` is implemented, I assume `IUSE="test"` should be added.

File: dev-util/emscripten/emscripten-1.34.11.ebuild [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+	net-libs/nodejs"
+DEPEND="${MYDEPEND}"
+RDEPEND="${MYDEPEND}"
+
+DEST="/usr/share/"
+TEST="${WORKDIR}/test/"
+
+src_prepare() {
+	cp "${FILESDIR}/emscripten.config" "${S}/"
+	cp "${FILESDIR}/99emscripten" "${S}/"
```
Comment (marbre) [inline dev-util/emscripten/emscripten-1.34.11.ebuild]: Missing die.

File: dev-util/emscripten/emscripten-1.34.11.ebuild [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+DESCRIPTION=" Emscripten: An LLVM-to-JavaScript Compiler"
+HOMEPAGE="http://emscripten.org/"
+SRC_URI="https://github.com/kripken/${PN}/archive/${PV}.tar.gz -> ${P}.tar.gz"
+KEYWORDS="~amd64 ~x86"
+SLOT="0"
+LICENSE="UoI-NCSA"
+
+MYDEPEND="${PYTHON_DEPS}
+	=dev-util/emscripten-fastcomp-1.34.11
+	net-libs/nodejs"
```
Comment (marbre) [inline dev-util/emscripten/emscripten-1.34.11.ebuild]: Why is `MYDEPEND` needed? Any reason not adding this directly to `DEPEND`?

File: dev-util/emscripten/emscripten-1.34.11.ebuild [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+	doenvd 99emscripten
+	ewarn "If you consider using emscripten in an active shell,"\
+		"please execute 'source /etc/profile'"
+}
+
+pkg_postinst() {
+	elog "Running emscripten initialization, may take a few seconds..."
+	export EM_CONFIG="${DEST}/${P}/emscripten.config" || die "Could not export variable"
+	/usr/bin/emcc -v || die "Could not run emcc initialization"
+}
```
Comment (marbre) [inline dev-util/emscripten/emscripten-1.34.11.ebuild]: `emscripten.config` is a python script, thus why is it needed to execute this and run emcc after installation?

File: dev-util/emscripten/emscripten-1.34.11.ebuild [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+src_prepare() {
+	cp "${FILESDIR}/emscripten.config" "${S}/"
+	cp "${FILESDIR}/99emscripten" "${S}/"
+}
+
+src_test() {
+	mkdir "${TEST}" || die "Could not create test directory!"
+	cp "${FILESDIR}/hello_world.cpp" "${TEST}" || die "Could not copy example file"
+	cp "${FILESDIR}/emscripten.config" "${TEST}" || die "Could not copy config file"
```
Comment (marbre) [inline dev-util/emscripten/emscripten-1.34.11.ebuild]: @jlec, maybe this should be moved to `src_prepare` as well?

File: dev-util/emscripten/emscripten-1.34.11.ebuild (hunk 1 above)
Comment (kiwifb) [inline dev-util/emscripten/emscripten-1.34.11.ebuild]: No, because the test flag is a "FEATURES" https://devmanual.gentoo.org/general-concepts/features/index.html .

Comment (marbre) [inline dev-util/emscripten/emscripten-1.34.11.ebuild]: You're right @kiwifb. Thanks for the hint.

File: dev-util/emscripten-fastcomp/emscripten-fastcomp-1.34.11.ebuild [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+inherit cmake-utils python-single-r1
+
+DESCRIPTION="Emscripten LLVM backend - Fastcomp is the default compiler core for Emscripten"
+HOMEPAGE="http://emscripten.org/"
+SRC_URI="https://github.com/kripken/${PN}/archive/${PV}.tar.gz -> ${P}.tar.gz
+	https://github.com/kripken/${PN}-clang/archive/${PV}.tar.gz -> ${PN}-clang-${PV}.tar.gz"
+KEYWORDS="~amd64 ~x86"
+SLOT="0"
+LICENSE="UoI-NCSA"
+
```
Comment (jlec) [inline dev-util/emscripten-fastcomp/emscripten-fastcomp-1.34.11.ebuild]: Missing

```ebuild
REQUIRED_USE="${PYTHON_REQUIRED_USE}"
```

File: dev-util/emscripten/emscripten-1.34.11.ebuild [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+# Distributed under the terms of the GNU General Public License v2
+# $Id$
+
+EAPI=5
+
+PYTHON_COMPAT=( python2_7 )
+
+inherit python-single-r1
+
+DESCRIPTION=" Emscripten: An LLVM-to-JavaScript Compiler"
```
Comment (jlec) [inline dev-util/emscripten/emscripten-1.34.11.ebuild]: unnecessary whitespace

File: dev-util/emscripten/emscripten-1.34.11.ebuild (hunk 3 above)
Comment (timeraider4u) [inline dev-util/emscripten/emscripten-1.34.11.ebuild]: Thx for pointing this out, DEPEND and RDEPEND were quite different before I cleaned them up so I just missed this one.

File: dev-util/emscripten/emscripten-1.34.11.ebuild (hunk 2 above)
Comment (timeraider4u) [inline dev-util/emscripten/emscripten-1.34.11.ebuild]: thx

File: dev-util/emscripten/emscripten-1.34.11.ebuild (hunk 4 above)
Comment (timeraider4u) [inline dev-util/emscripten/emscripten-1.34.11.ebuild]: emscripten.config stores all the paths. emcc reads this file (when the EM_CONFIG environment variable points to the file at the right location) and is needed to run the sanity checks at first run. When not run in pkg_postinst users will see warnings that they are not allowed to write to /usr/share/... EACH time emcc is called!

File: dev-util/emscripten-fastcomp/emscripten-fastcomp-1.34.11.ebuild (hunk 6 above)
Comment (timeraider4u) [inline dev-util/emscripten-fastcomp/emscripten-fastcomp-1.34.11.ebuild]: I just wanted to state that python-2.7 is required. Is it really required to add so many things or am I doing it wrong?

File: dev-util/emscripten/emscripten-1.34.11.ebuild (hunk 7 above)
Comment (timeraider4u) [inline dev-util/emscripten/emscripten-1.34.11.ebuild]: thx, a lot better.

File: dev-util/emscripten-fastcomp/emscripten-fastcomp-1.34.11.ebuild (hunk 6 above)
Comment (jlec) [inline dev-util/emscripten-fastcomp/emscripten-fastcomp-1.34.11.ebuild]: No it is. The line will force the user to have any python_target_*** in his USE to satisfy PYTHON_COMPAT. In you case there is only one choice. But the siutation changes when more ABIs are supported.

File: dev-util/emscripten/emscripten-1.34.11.ebuild (hunk 5 above)
Comment (jlec) [inline dev-util/emscripten/emscripten-1.34.11.ebuild]: either way works.

File: dev-util/emscripten-fastcomp/emscripten-fastcomp-1.34.11.ebuild (hunk 6 above)
Comment (timeraider4u) [inline dev-util/emscripten-fastcomp/emscripten-fastcomp-1.34.11.ebuild]: oke thx, good to know ;-).

**Comment by timeraider4u** [comment]
Thanks for your comments on the first version(s). I have now updated and re-committed the changed files. Feedback is welcome :-).

File: dev-util/emscripten-fastcomp/emscripten-fastcomp-1.34.11.ebuild (hunk 6 above)
Comment (jlec) [inline dev-util/emscripten-fastcomp/emscripten-fastcomp-1.34.11.ebuild]: Still missing.

**Comment by jlec** [comment]
Did you update the PR?

File: dev-util/emscripten-fastcomp/emscripten-fastcomp-1.34.11.ebuild (hunk 6 above)
Comment (timeraider4u) [inline dev-util/emscripten-fastcomp/emscripten-fastcomp-1.34.11.ebuild]: I see it here:
+EAPI=5
+
+PYTHON_COMPAT=( python2_7 )
+REQUIRED_USE="${PYTHON_REQUIRED_USE}"
+

Comment (jlec) [inline dev-util/emscripten-fastcomp/emscripten-fastcomp-1.34.11.ebuild]: ah true. please move this behind the IUSE line (or the place it would be). Always try to place as less as possible definitions before the inherit line. only things really used during global scope of the eclass need to go there.

Comment (timeraider4u) [inline dev-util/emscripten-fastcomp/emscripten-fastcomp-1.34.11.ebuild]: okay, thx, done :-).

File: dev-util/emscripten-fastcomp/emscripten-fastcomp-1.34.11.ebuild [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+REQUIRED_USE="${PYTHON_REQUIRED_USE}"
+
+DEPEND="${PYTHON_DEPEND}
```
Comment (jlec) [inline dev-util/emscripten-fastcomp/emscripten-fastcomp-1.34.11.ebuild]: Args, soryy but I just spotted it now. that should be ```${PYTHON_DEPS}```

Comment (timeraider4u) [inline dev-util/emscripten-fastcomp/emscripten-fastcomp-1.34.11.ebuild]: ah yes, thanks. Corrected by now.
